lua/entities/gmod_wire_expression2/core/extpp.lua

```diff
@@ -1,225 +1,146 @@
+	-- The 'n' type is actually 'normal'. Make an alias for number -> normal
 	preparsed_types = { ["NUMBER"] = "n" }
 end
 
+function E2Lib.ExtPP.Pass1(contents)
+	-- look for registerType lines and fill preparsed_types with them
+	for typename, typeid in string.gmatch("\n" .. contents, '%WregisterType%(%s*"(' .. p_typename .. ')"%s*,%s*"(' .. p_typeid .. ')"') do
+		preparsed_types[string.upper(typename)] = typeid
+	end
+end
+
 -- This function checks whether its argument is a valid type id.
 local function isValidTypeId(typeid)
 	return (typeid:match("^[a-wy-zA-WY-Z]$") or typeid:match("^[xX][a-wy-zA-WY-Z0-9][a-wy-zA-WY-Z0-9]$")) and true
 end
 
 -- Returns the typeid associated with the given typename
-local function getTypeId(typename)
-	local n = string.upper(typename)
-
-	-- was the type registered with E-2?
-	if wire_expression_types[n] then return wire_expression_types[n][1] end
-
-	-- was the name found when looking for registerType lines?
-	if preparsed_types[n] then return preparsed_types[n] end
+---@return string?
+local function getTypeId(ty)
+	local upper = ty:upper()
 
-	-- is the type name a valid typeid? use the type name as the typeid
-	if isValidTypeId(typename) then return typename end
+	return (wire_expression_types[upper] and wire_expression_types[upper][1]) -- Already registered.
+		or preparsed_types[upper]    -- Preparsed from registerType
+		or (isValidTypeId(ty) and ty) -- It is a type id. Weird.
 end
 
----@class ArgsKind
-local ArgsKind = {
-	None = 0,
-	Static = 1,
-	Variadic = 2,
-	VariadicTbl = 3
-}
-
--- parses an argument list
----@return { typeids: string[], argnames: string[] }, integer, string?
-local function parseArgs(args)
-	local argtable = { typeids = {}, argnames = {} }
-	if args:find("%S") == nil then return argtable, ArgsKind.None end -- no arguments
-
-	local args = args:Split(",")
-	local len = #args
-
-	for k, arg in ipairs(args) do
-		-- is this argument an ellipsis?
-		if arg:match( "^%s*%.%.%.%s*$") then
+-- Parses list of parameters
+---@param raw string
+---@param trace string
+---@return { [1]: string, [2]: string }[] parameters, boolean variadic, string? variadic_tbl
+local function parseParameters(raw, trace, trace_ext)
+	if not raw:match("%S") then return {}, false end
+
+	local parsed, split = {}, raw:Split(",")
+	local len = #split
+
+	for k, raw_param in ipairs(split) do
+		local name = raw_param:match("^%s*%.%.%.(%w+)%s*$")
+		if name then -- Variadic table parameter
+			assert(k == len, "PP syntax error: Ellipses table (..." .. name .. ") must be the last argument.")
+			return parsed, true, name
+		elseif raw_param:match("^%s*%.%.%.%s*$") then -- Variadic lua parameter
 			assert(k == len, "PP syntax error: Ellipses (...) must be the last argument.")
-			return argtable, ArgsKind.Variadic
+			ErrorNoHalt("Warning: Use of variadic parameter with ExtPP is not recommended and deprecated. Instead use ...<name> (which passes a table) or the `args` variable " .. trace_ext .. "\n")
```

Comment:
Why are we deviating from Lua's `...` here?

Reply:
This change is not part of the PR. https://github.com/wiremod/wire/pull/2372

I already explained why it's necessary there. tldr, doing that is essentially doing `{ unpack(args) }` which is just a massive waste of time (and is limited by lua's `unpack` limit). And deviating from lua really shouldn't be a concern since the extpp itself already breaks any lua tooling. At some point I will make an api without any sort of preprocessor

Reply:
Sure, the extra `unpack` is wasteful.
But the variadic functions we have aren't exactly called in a tight loop, so it's not like it matters.

We're not going to remove the Lua-style variadic syntax anyway, because that would break existing code.
And in some cases, this style is the most logical style, too, like in `printColor(...)`.
We should not throw errors at people for perfectly fine code.

Reply:
addressed in summary post => resolving